consensus/propagation/have_wants.go

```diff
@@ -37,25 +36,28 @@ func (blockProp *Reactor) handleHaves(peer p2p.ID, haves *proptypes.HaveParts, b
 		blockProp.Logger.Error("peer not found", "peer", peer)
 		return
 	}
+
 	_, parts, fullReqs, has := blockProp.getAllState(height, round)
 	if !has {
 		// TODO disconnect from the peer
 		blockProp.Logger.Error("received part state for unknown proposal", "peer", peer, "height", height, "round", round)
 		return
 	}
 
-	blockProp.mtx.RLock()
-	defer blockProp.mtx.RUnlock()
+	p.Initialize(height, round, int(parts.Total()))
 
-	// Update the peer's haves.
-	p.SetHaves(height, round, haves)
+	bm, _ := p.GetHaves(height, round)
 
-	if parts.IsComplete() {
+	for _, pmd := range haves.Parts {
+		bm.SetIndex(int(pmd.Index), true)
+	}
+
+	if parts.Original().IsComplete() {
```

Comment:
why we're only checking if the original is complete? maybe check if we have enough to decode the block?

Reply:
this is a good question, I think the thinking is this acts as a proxy for that, since as soon as we have enough parts to recover, we should have. it might be clearer to check for completeness of the combined partset

Reply:
makes sense